File: components/esp_board_manager/peripherals/periph_sdm/periph_sdm.c
/*
 * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
 * SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
 *
 * See LICENSE file for details.
 */

#include <stdlib.h>
#include "esp_log.h"
#include "periph_sdm.h"

static const char *TAG = "PERIPH_SDM";

int periph_sdm_init(void *cfg, int cfg_size, void **periph_handle)
{
    if (!cfg || !periph_handle || cfg_size != sizeof(sdm_config_t)) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }

    periph_sdm_handle_t *handle = (periph_sdm_handle_t *)calloc(1, sizeof(periph_sdm_handle_t));
    if (handle == NULL) {
        ESP_LOGE(TAG, "Failed to allocate memory for periph_sdm_handle_t");
        return -1;
    }

    sdm_config_t *sdm_cfg = (sdm_config_t *)cfg;
    esp_err_t err = sdm_new_channel(sdm_cfg, &handle->sdm_chan);
    if (err != ESP_OK) {
        free(handle);
        ESP_LOGE(TAG, "Call to sdm_new_channel failed: %s", esp_err_to_name(err));
        return -1;
    }

    ESP_LOGI(TAG, "SDM channel initialized successfully, GPIO: %d, sample rate: %" PRIu32 " Hz",
             sdm_cfg->gpio_num, sdm_cfg->sample_rate_hz);

    *periph_handle = handle;
    return 0;
}

int periph_sdm_deinit(void *periph_handle)
{
    if (!periph_handle) {
        ESP_LOGE(TAG, "Invalid handle");
        return -1;
    }

    periph_sdm_handle_t *handle = (periph_sdm_handle_t *)periph_handle;

    esp_err_t err = sdm_del_channel(handle->sdm_chan);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Call to sdm_del_channel failed: %s", esp_err_to_name(err));
        return -1;
    }

    free(handle);
    ESP_LOGI(TAG, "SDM channel deinitialized successfully");
    return 0;
}

File: components/esp_board_manager/peripherals/periph_sdm/periph_sdm.py
# SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
# SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
#
# See LICENSE file for details.

# SDM peripheral config parser
VERSION = 'v1.0.0'

PERIPH_SDM_IO_LIST = [
    'gpio_num',
]

import sys

import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'generators'))
from generators.utils.idf_version import get_idf_version

def get_includes() -> list:
    """Return list of required include headers for SDM peripheral"""
    return [
        'periph_sdm.h',
    ]

def parse(name: str, config: dict) -> dict:
    """Parse SDM peripheral configuration from YAML to C structure

    Args:
        name: Peripheral name
        config: Configuration dictionary from YAML

    Returns:
        Dictionary containing SDM configuration structure
    """
    try:
        # Get the config dictionary
        cfg = config.get('config', {})

        # Get GPIO pin number with validation
        gpio_num = cfg.get('gpio_num', -1)
        if gpio_num < 0:
            raise ValueError(f'Invalid GPIO pin number: {gpio_num}. Must be >= 0.')

        # Get clock source with validation
        clk_src = cfg.get('clk_src', 'SDM_CLK_SRC_DEFAULT')

        # Get sample rate with validation
        sample_rate_hz = cfg.get('sample_rate_hz', 1000000)
        if sample_rate_hz <= 0:
            raise ValueError(f'Invalid sample rate: {sample_rate_hz}. Must be > 0.')

        # Get invert_out flag with validation
        invert_out = cfg.get('invert_out', False)
        if not isinstance(invert_out, bool):
            raise ValueError(f'Invalid invert_out value: {invert_out}. Must be boolean.')

        # Get io_loop_back flag with validation
        io_loop_back = cfg.get('io_loop_back', False)
        if not isinstance(io_loop_back, bool):
            raise ValueError(f'Invalid io_loop_back value: {io_loop_back}. Must be boolean.')

        # Get allow_pd flag with validation
        allow_pd = cfg.get('allow_pd', False)
        if not isinstance(allow_pd, bool):
            raise ValueError(f'Invalid allow_pd value: {allow_pd}. Must be boolean.')

        flags = {
            'invert_out': 1 if invert_out else 0,
        }
        if get_idf_version()[0] < 6:
            flags['io_loop_back'] = 1 if io_loop_back else 0
            if allow_pd:
                print(f"YAML WARNING: 'allow_pd' flag in SDM peripheral '{name}' is not supported in IDF < 6.0 and will be ignored.")
        else:
            flags['allow_pd'] = 1 if allow_pd else 0
            if io_loop_back:
                print(f"YAML WARNING: 'io_loop_back' flag in SDM peripheral '{name}' is deprecated in IDF >= 6.0 and will be ignored.")

        # Create configuration structure
        result = {
            'struct_type': 'sdm_config_t',
            'struct_var': f'{name}_cfg',
            'struct_init': {
                'gpio_num': gpio_num,
                'clk_src': clk_src,
                'sample_rate_hz': sample_rate_hz,
                'flags': flags
            }
        }

        return result

    except ValueError as e:
        # Re-raise ValueError with more context
        raise ValueError(f"YAML validation error in SDM peripheral '{name}': {e}")
    except Exception as e:
        # Catch any other exceptions and provide context
        raise ValueError(f"Error parsing SDM peripheral '{name}': {e}")
